lib/krikri/ldp/invalidatable.rb

```diff
@@ -0,0 +1,102 @@
+module Krikri::LDP
+  ##
+  # Implements invalidation for `Krikri::LDP::Resource`s. This is different 
+  # from deletion, in that the resource continues to respond `200 OK`, and 
+  # return the representation, Nothing is removed from the LDP server.
+  # 
+  # Works as a mixin to `Krikri::LDP::Resource`, assuming an implementation of
+  # `#rdf_source`, which may simply return `self`.
+  #
+  # @example invalidating a resource
+  #   class MyResource
+  #     include Krikri::LDP::Resource
+  #     include Krikri::LDP::Invalidatable
+  #
+  #     def rdf_subject
+  #       @rdf_subject ||= RDF::URI('http://example.com/ldp/a/resource/path')
+  #     end
+  #   end
+  #  
+  #   invalidatable_resource = MyResource.new
+  #   # the resource must exist before it can be invalidated! 
+  #   invalidatable_resource.save
+  #
+  #   invalidatable_resource.invalidate!
+  #   invalidatable_resource.invalidated? # => true
+  #   invalidatable_resource.invalidated_at_time 
+  #   # => Thu, 03 Dec 2015 10:27:45 -0800
+  #
+  # @see http://www.w3.org/TR/2013/REC-prov-dm-20130430/#term-Invalidation 
+  #   for documentation on PROV invalidation
+  module Invalidatable
+    # @see RDF::PROV
+    INVALIDATED_BY_URI = RDF::PROV.wasInvalidatedBy
+    INVALIDATED_TIME_URI = RDF::PROV.invalidatedAtTime
+
+    ##
+    # Invalidates the resource by marking it with a `prov:invalidatedAtTime`. If
+    # an `RDF::Term` is passed as the first argument, that term is used as the 
+    # value of `prov:wasInvalidatedBy`.
+    #
+    # @example invalidating with an activity
+    #   invalidatable_resource.invalidate!(RDF::URI('http://example.org/moomin'))
+    #   invalidatable_resource.was_invalidated_by
+    #   # => #<RDF::URI:0x2acab846109c URI:http://example.org/moomin>
+    #
+    # @param activity_uri [RDF::Term] a URI for the invalidating activity. If 
+    #   none is given, this defaults to `nil` and no `prov:wasInvalidatedBy`
+    #   statement is added.
+    #
+    # @raise [RuntimeError] when the resource does not exist or is already 
+    #   invalid
+    # @return [void]
+    def invalidate!(activity_uri = nil)
+      raise "Cannot invalidate #{rdf_subject}, does not exist." unless exists?
+
+      # force a reload unless we have cached an invalidatedAtTime
+      rdf_source.get({}, true) unless invalidated?
+      # we check invalidated again in case the reload came back invalid
+      raise "Cannot invalidate #{rdf_subject}, already invalid." if invalidated?
```

Comment:
I considered that part of the contract of `#invalidate!` would be to update the timestamp and activity. So we should be able to reliably expect something like this to be true:

```ruby
aggregation.invalidate!(activity_uri)
aggregation.invalidated_at_time.second == DateTime.now.second
aggregation.invalidated_by == activity_uri
```

But since we don't want to change the timestamp or activity if the resource was already invalid, we need to throw an error.

The other option is to allow the `#invalidate!` to "fail" silently, but that seemed likely to bite us, to me. Thoughts?

For your use case, something like `aggregation.invalidate! if aggregation.valid?` seems workable. In general, I think we would want to be working with valid records only, in most cases.

All this said, I think if we keep things as is, we need to give this a better error class than `RuntimeError`.